python/lib/core/dmod/core/common/helper_functions.py

```diff
@@ -174,4 +183,323 @@ def merge_dictionaries(first: typing.Mapping = None, second: typing.Mapping = No
         if key_for_second not in merged_dictionary
     })
 
-    return merged_dictionary
\ No newline at end of file
+    return merged_dictionary
+
+
+def truncate(number: typing.Union[numbers.Number, float], digits: int) -> typing.Union[numbers.Number, float]:
+    """
+    Truncate a float by the desired number of digits
+
+    Example:
+        >>> example_value = 10.1231244121235435
+        >>> math.trunc(example_value)
+        10
+        >>> truncate(example_value, 3)
+        10.123
+        >>> truncate(example_value, 5)
+        10.12312
+
+    Args:
+        number: The number to truncate
+        digits: The number of desired digits
+
+    Returns:
+        The floating point number truncated to the specified number of digits
+    """
+    if digits <= 0:
+        return number
+
+    if not isinstance(digits, int):
+        raise TypeError(
+            f"The number {number} cannot be truncated - "
+            f"the number of digits to truncate by must be an integer and received {type(digits)} instead"
+        )
+
+    # We'll use the number 10.23423523423 and the desired number of digits as 3 as an example
+
+    # This gives us "10.23423523423"
+    string_representation = str(number)
+
+    if "." not in string_representation:
+        return number
+
+    # Find the number of current decimal points in order to avoid operations if necessary.
+    # This will yield 11 for our example
+    number_of_decimal_points = len(string_representation.split('.')[1])
+
+    # If the current number of digits is less than or equal to the number of desired digits, we can move on
+    # Our example has 11 decimal points, which is greater than the desired 3, so we continue
+    if number_of_decimal_points <= digits:
+        return number
+
+    # Create a factor that will shift the desired number of decimal points
+    # This will create 1000 for our example
+    adjuster = 10.0 ** digits
+
+    # This will shift the number by the desired number of decimal points
+    # This yields 10234.23523423 for our example
+    adjusted_value = adjuster * number
+
+    # This will get rid of the extra digits
+    # Our example value now becomes 10234.0
+    adjusted_value = math.trunc(adjusted_value)
+
+    # This will shift the value back to the desired number of decimal points
+    # Our example value now becomes 10.324
+    return adjusted_value / adjuster
+
+
+def get_subclasses(base: typing.Type[_CLASS_TYPE]) -> typing.List[typing.Type[_CLASS_TYPE]]:
+    """
+    Gets a collection of all concrete subclasses of the given class in memory
+
+    A subclass that has not been imported will not be returned
+
+    Example:
+        >>> import numpy
+        >>> get_subclasses(float)
+        [numpy.float64]
+
+    Args:
+        base: The base class to get subclasses from
+
+    Returns:
+        All implemented subclasses of a specified types
+    """
+    subclasses = [
```

Comment:
I take it you mean something along the lines of:

```python
    concrete_classes = list()

    for subclass in base.__subclasses__():
        if inspect.isabstract(subclass):
            concrete_classes.extend([
                cls
                for cls in get_subclasses(subclass)
                if cls not in concrete_classes
                   and not inspect.isabstract(cls)
            ])
        else:
            concrete_classes.append(subclass)

    return concrete_classes
```

The only reason I avoided it was because it's easier to stick your nose in it and investigate when debugging if the list to iterate is defined earlier. Due to lists of subclasses generally end up pretty short, I avoided the optimization of moving the first order filtering to remove the `if` condition from the for loop to make it as simple as possible.

There's not a huge difference, I'm not super attached to the different process, and the change has already been written above, so I can change it if you'd prefer.

Reply:
I'd also rather err on the side of clarity, so your comment is well taken! I was thinking you could just remove the `subclasses` list and replace `concrete_classes` with:

```python
concrete_classes = [
        subclass
        for subclass in base.__subclasses__()
        if not inspect.isabstract(subclass)
    ]
```

Like you mention, its not a big deal at all (the length of `subclasses` is never going to be very long). Ill leave this one up to you. If you want to make the nano-optimization, go for it, I think im just being a little nitpicky.

Reply:
I'm going with your suggestion since that initial `subclasses` collection is just a copy of `base.__subclasses__()` where we just need to iterate.

Reply:
Great, thanks for the dialog and for compromising!